src/main/java/com/artipie/asto/lock/StorageLock.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 artipie.com
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package com.artipie.asto.lock;
+
+import com.artipie.asto.Content;
+import com.artipie.asto.Key;
+import com.artipie.asto.Storage;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * {@link Lock} allowing to obtain lock on target {@link Key} in specified {@link Storage}.
+ * Lock is identified by it's unique identifier (UUID), which has to be different for each lock.
+ *
+ * @since 0.24
+ */
+public final class StorageLock implements Lock {
+
+    /**
+     * Storage.
+     */
+    private final Storage storage;
+
+    /**
+     * Target key.
+     */
+    private final Key target;
+
+    /**
+     * Identifier.
+     */
+    private final String uuid;
+
+    /**
+     * Ctor.
+     *
+     * @param storage Storage.
+     * @param target Target key.
+     */
+    public StorageLock(final Storage storage, final Key target) {
+        this(storage, target, UUID.randomUUID().toString());
```

Comment:
@olegmoz such implementation does not seem suitable for rpm: `Rpm#batchUpdate(com.artipie.asto.Key)` starts update rpm repository by given key and I need to be sure that this repo (by this key) is not updated simultaneously. It's not possible with `StorageLock` implementation as it will simply add new lock on each update request. Cashing `UUID <-> key` map is not a solution in case of several artipie instances working with single storage.

Reply:
@olenagerasimova `StorageLock` supposed to be created for each request and UUIDs have to be different for each request. When one lock is acquired (e.g. there is a value created from some UUID) next one will fail to acquire. This case is illustrated in test `shouldFailAcquireLockIfOtherProposalExists`.

Actually lock we have in RPM at the moment works the same way, see https://github.com/artipie/rpm-adapter/blob/master/src/main/java/com/artipie/rpm/misc/StorageLock.java#L83

Reply:
@olegmoz sorry, I've read the code not very carefully